gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/structure/io/binary/types/MapSerializer.java

```diff
@@ -24,20 +24,37 @@
 import org.apache.tinkerpop.gremlin.structure.io.Buffer;
 
 import java.io.IOException;
+import java.util.HashMap;
 import java.util.LinkedHashMap;
 import java.util.Map;
 import java.util.Set;
 
 public class MapSerializer extends SimpleTypeSerializer<Map> {
+
+    private byte valueFlag = (byte)0x00;
+
     public MapSerializer() {
         super(DataType.MAP);
     }
 
+    @Override
+    public Map readValue(final Buffer buffer, final GraphBinaryReader context, final boolean nullable) throws IOException {
+        if (nullable) {
+            valueFlag = buffer.readByte();
+            if ((valueFlag & 1) == 1) {
+                return null;
+            }
+        }
+
+        return readValue(buffer, context);
+    }
+
     @Override
     protected Map readValue(final Buffer buffer, final GraphBinaryReader context) throws IOException {
         final int length = buffer.readInt();
 
-        final Map<Object,Object> result = new LinkedHashMap<>(length);
+        // original read value outputs LinkedHashMap, do we want to keep all result as LinkedHashMap anyway?
```

Comment:
This question does have me wondering if we've landed on the right design with this ordered map flag. In theory the only downside of using LinkedHashMap all the time is a slight increase in overhead to maintain the links. Considering Java has always converted all maps to LinkedHashMap upon deserialization, and that JS and Python both implicitly deserialize to ordered maps, I'm wondering if the best answer is to update our spec to state "All gremlin language drivers should attempt to deserialize to an ordered map wherever possible".

As far as the official GLVs go, this would require a switch in .Net to using `OrderedDictionary`, whereas Go would need to have lack of ordering added as a documented limitation.